File: backend/mr/src/main/java/io/featurehub/mr/resources/PortfolioResource.java
package io.featurehub.mr.resources;

import io.featurehub.db.api.FillOpts;
import io.featurehub.db.api.GroupApi;
import io.featurehub.db.api.OptimisticLockingException;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.OrganizationApi;
import io.featurehub.db.api.PortfolioApi;
import io.featurehub.mr.api.PortfolioServiceDelegate;
import io.featurehub.mr.auth.AuthManagerService;
import io.featurehub.mr.model.ApplicationGroupRole;
import io.featurehub.mr.model.ApplicationRoleType;
import io.featurehub.mr.model.Group;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.Portfolio;
import io.featurehub.mr.utils.PortfolioUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.ForbiddenException;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class PortfolioResource implements PortfolioServiceDelegate {
  private static final Logger log = LoggerFactory.getLogger(PortfolioResource.class);

  private final GroupApi groupApi;
  private final AuthManagerService authManager;
  private final PortfolioApi portfolioApi;
  private final OrganizationApi organizationApi;
  private final PortfolioUtils portfolioUtils;

  @Inject
  public PortfolioResource(GroupApi groupApi, AuthManagerService authManager, PortfolioApi portfolioApi, OrganizationApi organizationApi, PortfolioUtils portfolioUtils) {
    this.groupApi = groupApi;
    this.authManager = authManager;
    this.portfolioApi = portfolioApi;
    this.organizationApi = organizationApi;
    this.portfolioUtils = portfolioUtils;
  }

  @Override
  public Portfolio createPortfolio(Portfolio portfolio, CreatePortfolioHolder holder, SecurityContext securityContext) {
    if (authManager.isOrgAdmin(authManager.from(securityContext))) {
      Opts opts = new Opts().add(FillOpts.Groups, holder.includeGroups).add(FillOpts.Applications, holder.includeApplications);

      Portfolio created;
      try {
        created = portfolioApi.createPortfolio(portfolio, opts, authManager.from(securityContext));
      } catch (PortfolioApi.DuplicatePortfolioException e) {
        log.error("Duplicate portfolio name", e);
        throw new WebApplicationException(Response.Status.CONFLICT);
      }

      if (created == null) {
        throw new NotFoundException("No conditions to allow you to create.");
      }

      Group group;
      try {
        group = groupApi.createPortfolioGroup(created.getId(),
          new Group()
            .name(portfolioUtils.formatPortfolioAdminGroupName(portfolio))
            .admin(true)
            .portfolioId(created.getId()),
          authManager.from(securityContext));
      } catch (GroupApi.DuplicateGroupException e) {
        throw new WebApplicationException(Response.Status.CONFLICT);
      }

      created.addGroupsItem(group);

      return created;
    }

    throw new ForbiddenException("No permission to create portfolio");
  }

  @Override
  public Boolean deletePortfolio(String id, DeletePortfolioHolder holder, SecurityContext securityContext) {
    final Person from = authManager.from(securityContext);
    if (authManager.isOrgAdmin(from)) {
      if (portfolioApi.getPortfolio(id, Opts.empty(), from) != null) {
        portfolioApi.deletePortfolio(id);
        return true;
      } else {
        return false; // no portfolio
      }
    }

    throw new ForbiddenException("Not allowed to delete portfolio");
  }

  @Override
  public List<Portfolio> findPortfolios(FindPortfoliosHolder holder, SecurityContext securityContext) {
    return portfolioApi.findPortfolios(holder.filter,
      authManager.orgPersonIn(
        authManager.from(securityContext)
      ), holder.order,
      new Opts().add(FillOpts.Groups, holder.includeGroups).add(FillOpts.Applications, holder.includeApplications),
      authManager.from(securityContext));
  }

  @Override
  public Portfolio getPortfolio(String id, GetPortfolioHolder holder, SecurityContext securityContext) {
    Portfolio portfolio = portfolioApi.getPortfolio(id,
      new Opts().add(FillOpts.Groups, holder.includeGroups)
        .add(FillOpts.Applications, holder.includeApplications)
        .add(FillOpts.Environments, holder.includeEnvironments)
      , authManager.from(securityContext));
    if (portfolio == null) {
      throw new NotFoundException("No such portfolio");
    }

    return portfolio;
  }

  @Override
  public Portfolio updatePortfolio(String id, Portfolio portfolio, UpdatePortfolioHolder holder, SecurityContext securityContext) {
    Person current = authManager.from(securityContext);

    if (authManager.isPortfolioAdmin(id, current, null)) {
      portfolio.setId(id);
      Portfolio pf;

      try {
        pf = portfolioApi.updatePortfolio(portfolio, Opts.empty());
      } catch (PortfolioApi.DuplicatePortfolioException e) {
        log.error("Duplicate portfolio name", e);
        throw new WebApplicationException(Response.Status.CONFLICT);
      } catch (OptimisticLockingException e) {
        throw new WebApplicationException(422);
      }

      if (pf == null) {
        throw new NotFoundException("No such portfolio");
      }

      groupApi.updateAdminGroupForPortfolio(pf.getId(), portfolioUtils.formatPortfolioAdminGroupName(pf));

      return portfolioApi.getPortfolio(pf.getId(), new Opts().add(FillOpts.Groups, holder.includeGroups)
        .add(FillOpts.Applications, holder.includeApplications)
        .add(FillOpts.Environments, holder.includeEnvironments), current);
    }

    throw new ForbiddenException("Not an admin, cannot rename portfolio");
  }
}

File: backend/mr/src/main/kotlin/io/featurehub/mr/resources/PortfolioResource.kt
package io.featurehub.mr.resources

import io.featurehub.db.api.*
import io.featurehub.mr.api.PortfolioServiceDelegate
import io.featurehub.mr.auth.AuthManagerService
import io.featurehub.mr.model.CreateGroup
import io.featurehub.mr.model.CreatePortfolio
import io.featurehub.mr.model.Portfolio
import io.featurehub.mr.utils.PortfolioUtils
import jakarta.inject.Inject
import jakarta.ws.rs.ForbiddenException
import jakarta.ws.rs.NotFoundException
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.SecurityContext
import org.slf4j.LoggerFactory
import java.util.*

class PortfolioResource @Inject constructor(
  private val groupApi: GroupApi,
  private val authManager: AuthManagerService,
  private val portfolioApi: PortfolioApi,
  private val organizationApi: OrganizationApi,
  private val portfolioUtils: PortfolioUtils
) : PortfolioServiceDelegate {

  override fun createPortfolio(
    createPortfolio: CreatePortfolio,
    holder: PortfolioServiceDelegate.CreatePortfolioHolder,
    securityContext: SecurityContext?
  ): Portfolio {
    if (authManager.isOrgAdmin(authManager.from(securityContext))) {
      val opts =
        Opts().add(FillOpts.Groups, holder.includeGroups).add(FillOpts.Applications, holder.includeApplications)
      val created = try {
        portfolioApi.createPortfolio(createPortfolio, opts, authManager.who(securityContext)) ?: throw NotFoundException()
      } catch (e: PortfolioApi.DuplicatePortfolioException) {
        log.error("Duplicate portfolio name", e)
        throw WebApplicationException(Response.Status.CONFLICT)
      }
      val group = try {
        groupApi.createGroup(
          created.id,
          CreateGroup()
            .name(portfolioUtils.formatPortfolioAdminGroupName(createPortfolio.name))
            .admin(true),
          authManager.from(securityContext)
        ) ?: throw NotFoundException()
      } catch (e: GroupApi.DuplicateGroupException) {
        throw WebApplicationException(Response.Status.CONFLICT)
      }
      created.addGroupsItem(group)
      return created
    }
    throw ForbiddenException("No permission to create portfolio")
  }

  override fun deletePortfolio(
    id: UUID,
    holder: PortfolioServiceDelegate.DeletePortfolioHolder,
    securityContext: SecurityContext
  ): Boolean {
    val from = authManager.who(securityContext)
    if (authManager.isOrgAdmin(from)) {
      return if (portfolioApi.getPortfolio(id, Opts.empty(), from) != null) {
        portfolioApi.deletePortfolio(id)
        true
      } else {
        false // no portfolio
      }
    }
    throw ForbiddenException("Not allowed to delete portfolio")
  }

  override fun findPortfolios(
    holder: PortfolioServiceDelegate.FindPortfoliosHolder,
    securityContext: SecurityContext
  ): List<Portfolio> {
    return portfolioApi.findPortfolios(
      holder.filter,
      holder.order,
      Opts().add(FillOpts.Groups, holder.includeGroups).add(FillOpts.Applications, holder.includeApplications),
      authManager.who(securityContext)
    )
  }

  override fun getPortfolio(
    id: UUID,
    holder: PortfolioServiceDelegate.GetPortfolioHolder,
    securityContext: SecurityContext
  ): Portfolio {
    return portfolioApi.getPortfolio(
      id,
      Opts().add(FillOpts.Groups, holder.includeGroups)
        .add(FillOpts.Applications, holder.includeApplications)
        .add(FillOpts.Environments, holder.includeEnvironments), authManager.who(securityContext)
    ) ?: throw NotFoundException("No such portfolio")
  }


  override fun updatePortfolioOnOrganisation(
    portfolio: Portfolio,
    holder: PortfolioServiceDelegate.UpdatePortfolioOnOrganisationHolder,
    securityContext: SecurityContext?
  ): Portfolio {
    val current = authManager.from(securityContext)

    if (authManager.isPortfolioAdmin(portfolio.id, current, null)) {
      val pf = try {
        portfolioApi.updatePortfolio(portfolio, Opts.empty()) ?: throw NotFoundException("No such portfolio")
      } catch (e: PortfolioApi.DuplicatePortfolioException) {
        log.error("Duplicate portfolio name", e)
        throw WebApplicationException(Response.Status.CONFLICT)
      } catch (e: OptimisticLockingException) {
        throw WebApplicationException(422)
      }

      groupApi.updateAdminGroupForPortfolio(pf.id, portfolioUtils.formatPortfolioAdminGroupName(pf.name))
      return portfolioApi.getPortfolio(
        pf.id, Opts().add(FillOpts.Groups, holder.includeGroups)
          .add(FillOpts.Applications, holder.includeApplications)
          .add(FillOpts.Environments, holder.includeEnvironments), current.id!!.id
      ) ?: throw NotFoundException()
    }

    throw ForbiddenException("Not an admin, cannot rename portfolio")
  }

  @Deprecated("Deprecated in Java")
  override fun updatePortfolio(
    id: UUID, portfolio: Portfolio, holder: PortfolioServiceDelegate.UpdatePortfolioHolder,
    securityContext: SecurityContext
  ): Portfolio {
    val current = authManager.from(securityContext)
    if (authManager.isPortfolioAdmin(id, current, null)) {
      portfolio.id = id
      val pf: Portfolio?
      pf = try {
        portfolioApi.updatePortfolio(portfolio, Opts.empty())
      } catch (e: PortfolioApi.DuplicatePortfolioException) {
        log.error("Duplicate portfolio name", e)
        throw WebApplicationException(Response.Status.CONFLICT)
      } catch (e: OptimisticLockingException) {
        throw WebApplicationException(422)
      }
      if (pf == null) {
        throw NotFoundException("No such portfolio")
      }
      groupApi.updateAdminGroupForPortfolio(pf.id, portfolioUtils.formatPortfolioAdminGroupName(pf.name))
      return portfolioApi.getPortfolio(
        pf.id, Opts().add(FillOpts.Groups, holder.includeGroups)
          .add(FillOpts.Applications, holder.includeApplications)
          .add(FillOpts.Environments, holder.includeEnvironments), current.id!!.id
      ) ?: throw NotFoundException()
    }
    throw ForbiddenException("Not an admin, cannot rename portfolio")
  }

  companion object {
    private val log = LoggerFactory.getLogger(PortfolioResource::class.java)
  }
}
